bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java

```diff
@@ -0,0 +1,546 @@
+    private static final String HUB_TOKEN_ZONE_LIST = "ZoneList";
+    private static final String HUB_SERVICE_REPORT_ZONE_PROPERTIES = "ReportZoneProperties";
+    private static final String HUB_SERVICE_ZONE_PROPERTIES_CHANGED = "ZonePropertiesChanged";
+    private static final String HUB_SERVICE_LIST_ZONE = "ListZones";
+    private static final String HUB_SERVICE_SYSTEM_INFO = "SystemInfo";
+
+    private @Nullable Future<?> hubController;
+    private String hubHost;
+    private int hubPort;
+    private @Nullable Socket hubSocket;
+    private @Nullable PrintStream hubOut;
+    private @Nullable Scanner hubIn;
+    private CompletableFuture<@Nullable Void> hubControllerConnected;
+    private int hubReconnectSleep; // Sleep time before we attempt re-connect
+    private ScheduledExecutorService scheduler;
+
+    private AtomicBoolean stop; // Stop the controller as soon as possible
+    private AtomicBoolean stopWhenCommandsServed; // Stop the controller once all pending commands have been served
+    private long stopTimestamp = 0; // Stop the controller after this timestamp
+
+    // When we submit commmands to the hub we don't correlate commands and responses. We simply use the first available
+    // response that answers our question. For that we maintain lists of all pending commands.
+    private List<AdorneHubController.ZoneIDFutureRecord> stateCommands;
+    private List<CompletableFuture<List<Integer>>> zoneCommands;
+    private List<CompletableFuture<String>> macAddressCommands;
+    private int commandId; // We assign increasing command ids to all rest commands to the hub for easier
+                           // troubleshooting
+
+    private @Nullable AdorneHubChangeNotify changeListener;
+    private Gson gson;
+
+    private static class ZoneIDFutureRecord {
+        public int zoneId;
+        public CompletableFuture<AdorneDeviceState> future;
+
+        ZoneIDFutureRecord(int zoneId, CompletableFuture<AdorneDeviceState> future) {
+            this.zoneId = zoneId;
+            this.future = future;
+        }
+    }
+
+    public AdorneHubController(@Nullable String hubHost, @Nullable Integer hubPort,
+            ScheduledExecutorService scheduler) {
+        hubController = null;
+        this.hubHost = (hubHost == null) ? HUB_DEFAULT_HOST : hubHost;
+        this.hubPort = hubPort == null ? HUB_DEFAULT_PORT : hubPort;
+        hubSocket = null;
+        hubOut = null;
+        hubIn = null;
+        hubControllerConnected = new CompletableFuture<>();
+        ;
+        this.scheduler = scheduler;
+
+        stop = new AtomicBoolean(false);
+        stopWhenCommandsServed = new AtomicBoolean(false);
+        stopTimestamp = 0;
+
+        stateCommands = new ArrayList<>();
+        zoneCommands = new ArrayList<>();
+        macAddressCommands = new ArrayList<>();
+        commandId = 0;
+
+        changeListener = null;
+        gson = new Gson();
```

Comment:
Do you need a new Gson-instance or could you share one from the handler factory and pass it though the handler's constructor to the controller?

Reply:
Yes, I could reuse the gson instance. But using Gson is an implementation detail of the hub controller that the factory doesn't need to know about. I could just lazy load the gson initialization into a static field in the controller constructor. That way I get re-use while containing logic in the controller.
That said, I did a quick perf test and the initialization takes less than 1ms on my RaspberryPi (since I am not using any custom JSON models overhead is minimal) . Also, savings would only apply to the edge case of running multiple hubs. Is it worth adding code complexity for this optimization?

Reply:
If it is unlikely that someone uses more than one hub, I'm ok with the way it is.